torchtune/models/qwen2_5/_tokenizer.py

```diff
@@ -0,0 +1,232 @@
+                sequence of messages. Default is True.
+
+        Returns:
+            Tuple[List[int], List[bool]]: The list of token ids and the list of masks.
+
+        Raises:
+            RuntimeError: If a message contains non-text content
+        """
+        assert not isinstance(
+            self.prompt_template, ChatMLTemplate
+        ), "Using ChatMLTemplate with tokenize_messages will result in multiple <|im_*|> tokens wrapping each message."
+        templated_messages = (
+            self.prompt_template(messages)
+            if self.prompt_template is not None
+            else messages
+        )
+
+        tokenized_messages = []
+        mask = []
+        for i, message in enumerate(templated_messages):
+            tokens = []
+
+            # message header
+            if message.role == "ipython":
+                if i == 0 or templated_messages[i - 1].role != "ipython":
+                    self._add_message_start_tokens(tokens, "user")
+                    tokens.extend(
+                        self.encode("<tool_response>\n", add_bos=False, add_eos=False)
+                    )
+                else:
+                    tokens.extend(
+                        self.encode("\n<tool_response>\n", add_bos=False, add_eos=False)
+                    )
+            else:
+                self._add_message_start_tokens(tokens, message.role)
+                if message.role == "assistant" and message.ipython:
+                    tokens.extend(
+                        self.encode("<tool_call>\n", add_bos=False, add_eos=False)
+                    )
+
+            # message content
+            for item in message.content:
+                if item["type"] == "text":
+                    tokens.extend(
+                        self.encode(
+                            item["content"],
+                            add_bos=False,
+                            add_eos=False,
+                        )
+                    )
+                else:
+                    raise RuntimeError(
+                        f"Unsupported message content type: {item['type']}"
+                    )
+
+            # message footer
+            if message.role == "ipython":
+                if (
+                    i == len(templated_messages) - 1
+                    or templated_messages[i + 1].role != "ipython"
+                ):
+                    tokens.extend(
+                        self.encode("\n</tool_response>", add_bos=False, add_eos=False)
+                    )
+                    self._add_message_end_tokens(tokens)
+                else:
+                    tokens.extend(
+                        self.encode("\n</tool_response>", add_bos=False, add_eos=False)
+                    )
+            else:
+                if message.role == "assistant" and message.ipython:
+                    tokens.extend(
+                        self.encode("\n</tool_call>", add_bos=False, add_eos=False)
```

Comment:
would just add the id directly since you defined it in the special tokens dict above. or better yet I would save the tool call id as an attribute and add it directly here instead of encoding in case the user wants to change the id:
```
self.tool_call_start_id = self.special_tokens["<tool_call>"]
self.tool_call_end_id = self.special_tokens["</tool_call>"]
...
tokens.extend([self.tool_call_start_id])
```
so you don't have to encode every single time, save a tiny bit of time. same comment for tool response.

I wonder how a user would go about changing the actual encoded string for tool calls though, I guess there's no easy way to do that...